srml/babe/src/lib.rs

```diff
@@ -127,25 +137,68 @@ impl<T: Trait> Module<T> {
 		// the majority of their slot.
 		<timestamp::Module<T>>::minimum_period().saturating_mul(2.into())
 	}
-}
-
-impl<T: Trait> OnTimestampSet<T::Moment> for Module<T> {
-	fn on_timestamp_set(_moment: T::Moment) { }
-}
 
-impl<T: Trait> Module<T> {
 	fn change_authorities(new: Vec<AuthorityId>) {
 		<Authorities<T>>::put(&new);
 
 		let log: DigestItem<T::Hash> = DigestItem::Consensus(BABE_ENGINE_ID, new.encode());
 		<system::Module<T>>::deposit_log(log.into());
 	}
+
+	fn deposit_vrf_output(vrf_output: &[u8; VRF_OUTPUT_LENGTH]) {
+		let l = <Randomness<T>>::get();
+		let mut arr = [0u8; VRF_OUTPUT_LENGTH + 32];
+		arr[0..32].copy_from_slice(&l[..]);
+		arr[32..VRF_OUTPUT_LENGTH + 32].copy_from_slice(&vrf_output[..]);
+		<Randomness<T>>::put(runtime_io::blake2_256(&arr));
+	}
+
+	pub fn process_inherent_digests() {
+		let mut is_first_babe_digest = true;
+		for i in Self::get_inherent_digests()
+			.logs
+			.iter()
+			.filter_map(|s| s.as_pre_runtime())
+			.filter_map(|(engine, mut data)| if engine == BABE_ENGINE_ID {
+				Decode::decode(&mut data)
+			} else { None }) {
+			assert!(is_first_babe_digest, "BABE only allows one BABE pre-digest; qed");
+			is_first_babe_digest = false;
+			let (ref vrf_output, ref _vrf_proof, ref _author, _slot_num): (
+				[u8; VRF_OUTPUT_LENGTH],
+				[u8; VRF_PROOF_LENGTH],
+				[u8; PUBLIC_KEY_LENGTH],
+				u64,
+			) = i;
+			Self::deposit_vrf_output(vrf_output);
+		}
+		assert!(!is_first_babe_digest, "BABE requires exactly one BABE pre-digest; qed")
+	}
+
+	fn get_inherent_digests() -> system::DigestOf<T> {
+		<system::Module<T>>::get_inherent_digests()
+	}
+
+	#[allow(unsafe_code)]
+	fn hash_randomness(vrf_outputs: Vec<[u8; VRF_OUTPUT_LENGTH]>) -> [u8; VRF_OUTPUT_LENGTH] {
+		// yucky unsafe code, but at least I don’t need to write 32 XORs!
```

Comment:
why not just `xor` in a `zip`? rustc should be able to generate decent assembly

Reply:
https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=afcb3c115bd1ad19e2c72981ded96dbd

`zip` ASM:

```
playground::xor_in_u8:
	testq	%rdx, %rdx
	je	.LBB7_1
	shlq	$5, %rdx
	leaq	-32(%rdx), %r8
	movl	%r8d, %ecx
	shrl	$5, %ecx
	addl	$1, %ecx
	andq	$3, %rcx
	je	.LBB7_3
	negq	%rcx
	xorps	%xmm0, %xmm0
	movq	%rsi, %rax
	xorps	%xmm1, %xmm1

.LBB7_5:
	movups	(%rax), %xmm2
	xorps	%xmm2, %xmm0
	movups	16(%rax), %xmm2
	leaq	32(%rax), %rax
	xorps	%xmm2, %xmm1
	incq	%rcx
	jne	.LBB7_5
	cmpq	$96, %r8
	jae	.LBB7_7
	jmp	.LBB7_9

.LBB7_1:
	xorps	%xmm0, %xmm0
	xorps	%xmm1, %xmm1
	movups	%xmm0, (%rdi)
	movups	%xmm1, 16(%rdi)
	retq

.LBB7_3:
	xorps	%xmm0, %xmm0
	movq	%rsi, %rax
	xorps	%xmm1, %xmm1
	cmpq	$96, %r8
	jb	.LBB7_9

.LBB7_7:
	addq	%rdx, %rsi

.LBB7_8:
	movups	(%rax), %xmm2
	xorps	%xmm0, %xmm2
	movups	16(%rax), %xmm0
	xorps	%xmm1, %xmm0
	movups	32(%rax), %xmm1
	movups	48(%rax), %xmm3
	movups	64(%rax), %xmm4
	xorps	%xmm1, %xmm4
	xorps	%xmm2, %xmm4
	movups	80(%rax), %xmm2
	xorps	%xmm3, %xmm2
	xorps	%xmm0, %xmm2
	movups	96(%rax), %xmm0
	xorps	%xmm4, %xmm0
	movups	112(%rax), %xmm1
	xorps	%xmm2, %xmm1
	subq	$-128, %rax
	cmpq	%rsi, %rax
	jne	.LBB7_8

.LBB7_9:
	movups	%xmm0, (%rdi)
	movups	%xmm1, 16(%rdi)
	retq
```

unsafe-code ASM
```
playground::xor_in_u64:
	testq	%rdx, %rdx
	je	.LBB8_1
	shlq	$5, %rdx
	leaq	-32(%rdx), %r8
	movl	%r8d, %ecx
	shrl	$5, %ecx
	addl	$1, %ecx
	andq	$3, %rcx
	je	.LBB8_3
	negq	%rcx
	xorps	%xmm0, %xmm0
	movq	%rsi, %rax
	xorps	%xmm1, %xmm1

.LBB8_5:
	movups	(%rax), %xmm2
	xorps	%xmm2, %xmm0
	movups	16(%rax), %xmm2
	leaq	32(%rax), %rax
	xorps	%xmm2, %xmm1
	incq	%rcx
	jne	.LBB8_5
	cmpq	$96, %r8
	jae	.LBB8_7
	jmp	.LBB8_9

.LBB8_1:
	xorps	%xmm0, %xmm0
	xorps	%xmm1, %xmm1
	movups	%xmm0, (%rdi)
	movups	%xmm1, 16(%rdi)
	retq

.LBB8_3:
	xorps	%xmm0, %xmm0
	movq	%rsi, %rax
	xorps	%xmm1, %xmm1
	cmpq	$96, %r8
	jb	.LBB8_9

.LBB8_7:
	addq	%rdx, %rsi

.LBB8_8:
	movups	(%rax), %xmm2
	xorps	%xmm0, %xmm2
	movups	16(%rax), %xmm0
	xorps	%xmm1, %xmm0
	movups	32(%rax), %xmm1
	movups	48(%rax), %xmm3
	movups	64(%rax), %xmm4
	xorps	%xmm1, %xmm4
	xorps	%xmm2, %xmm4
	movups	80(%rax), %xmm2
	xorps	%xmm3, %xmm2
	xorps	%xmm0, %xmm2
	movups	96(%rax), %xmm0
	xorps	%xmm4, %xmm0
	movups	112(%rax), %xmm1
	xorps	%xmm2, %xmm1
	subq	$-128, %rax
	cmpq	%rsi, %rax
	jne	.LBB8_8

.LBB8_9:
	movups	%xmm0, (%rdi)
	movups	%xmm1, 16(%rdi)
	retq

.LCPI9_0:
	.quad	255
	.quad	255
```

Reply:
As an aside, I'd kinda think `#[repr(align(8))]` on a wrapper type might fix any performance hit one does find in such situations, not sure.

Reply:
The function that used the unsafe code is now gone.